daemon/sync.go

```diff
@@ -0,0 +1,399 @@
+			eventTypes[event.EventRelease] = true
+		case update.Images:
+			spec := n.Spec.Spec.(update.ReleaseImageSpec)
+			noteEvents = append(noteEvents, event.Event{
+				ServiceIDs: n.Result.AffectedResources(),
+				Type:       event.EventRelease,
+				StartedAt:  s.started,
+				EndedAt:    time.Now().UTC(),
+				LogLevel:   event.LogLevelInfo,
+				Metadata: &event.ReleaseEventMetadata{
+					ReleaseEventCommon: event.ReleaseEventCommon{
+						Revision: c.commits[i].Revision,
+						Result:   n.Result,
+						Error:    n.Result.Error(),
+					},
+					Spec: event.ReleaseSpec{
+						Type:             event.ReleaseImageSpecType,
+						ReleaseImageSpec: &spec,
+					},
+					Cause: n.Spec.Cause,
+				},
+			})
+			eventTypes[event.EventRelease] = true
+		case update.Auto:
+			spec := n.Spec.Spec.(update.Automated)
+			noteEvents = append(noteEvents, event.Event{
+				ServiceIDs: n.Result.AffectedResources(),
+				Type:       event.EventAutoRelease,
+				StartedAt:  s.started,
+				EndedAt:    time.Now().UTC(),
+				LogLevel:   event.LogLevelInfo,
+				Metadata: &event.AutoReleaseEventMetadata{
+					ReleaseEventCommon: event.ReleaseEventCommon{
+						Revision: c.commits[i].Revision,
+						Result:   n.Result,
+						Error:    n.Result.Error(),
+					},
+					Spec: spec,
+				},
+			})
+			eventTypes[event.EventAutoRelease] = true
+		case update.Policy:
+			// Use this to mean any change to policy
+			eventTypes[event.EventUpdatePolicy] = true
+		default:
+			// Presume it's not something we're otherwise sending
+			// as an event
+			eventTypes[event.NoneOfTheAbove] = true
+		}
+	}
+	return noteEvents, eventTypes, nil
+}
+
+// logCommitEvent reports all synced commits to the upstream.
+func logCommitEvent(s *Sync, c changeset, serviceIDs flux.ResourceIDSet,
+	includesEvents map[string]bool, resourceErrors []event.ResourceError) error {
+	cs := make([]event.Commit, len(c.commits))
+	for i, ci := range c.commits {
+		cs[i].Revision = ci.Revision
+		cs[i].Message = ci.Message
+	}
+	if err := s.LogEvent(event.Event{
+		ServiceIDs: serviceIDs.ToSlice(),
+		Type:       event.EventSync,
+		StartedAt:  s.started,
+		EndedAt:    s.started,
+		LogLevel:   event.LogLevelInfo,
+		Metadata: &event.SyncEventMetadata{
+			Commits:     cs,
+			InitialSync: c.initialSync,
+			Includes:    includesEvents,
+			Errors:      resourceErrors,
+		},
+	}); err != nil {
+		s.logger.Log("err", err)
+		return err
+	}
+	return nil
+}
+
+// moveSyncTag moves the sync tag to the revision we just synced.
+func moveSyncTag(ctx context.Context, s *Sync, c changeset) error {
+	tagAction := git.TagAction{
+		Revision: c.newTagRev,
+		Message:  "Sync pointer",
+	}
+	ctx, cancel := context.WithTimeout(ctx, s.gitConfig.Timeout)
+	if err := s.working.MoveSyncTagAndPush(ctx, tagAction); err != nil {
+		return err
+	}
+	cancel()
+	return nil
+}
+
+// refresh refreshes the repository, notifying the daemon we have a new
+// sync head.
+func refresh(ctx context.Context, s *Sync) error {
```

Comment:
Is this used? I couldn't find where

Reply:
What IDE are you using? GoLand grays out  orphan private symbols. I highly recommend it. (Sorry for the not so useful comment, I am AFK and can't check it right now).

Reply:
Emacs. Is GoLand also an NNTP client? I think not.

Reply:
/me drops his case

Reply:
/offtopic

I have started trying out GoLand again (used GogLand in the past when it was still slow and unreliable) and it is much better than VSC.

Reply:
It is used here: https://github.com/weaveworks/flux/pull/1791/files/4536f7840682032c19bbf07b5ba6403ef941fa80#diff-48e0783dedd59969ac55a5997f8c2ee9R125